### PR TITLE
Add sentry

### PR DESCRIPTION
**Pull request by iAdramelk**
Test integration with Sentry. To make it work we need to add env variable `SENTRY_DSN` to Heroku.

### CODE REVIEW

**Comment by jorgeorpinel** [comment]
Interesting. So Sentry is a SRE/APM dashboard? I quickly read https://blog.sentry.io/2018/03/06/the-sentry-workflow but would be great to know what this is all about 🙂

**Comment by iAdramelk** [comment]
@jorgeorpinel , yep. Pretty much it. We can collect info about user's javascirpt errors with it, see information about environment, stack traces, etc.

File: pages/_app.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import * as Sentry from '@sentry/browser'
+
+Sentry.init({
+  dsn: process.env.SENTRY_DSN
+})
+
+class MyApp extends App {
+  static async getInitialProps({ Component, ctx }) {
+    let pageProps = {}
+
```
Comment (jorgeorpinel) [inline pages/_app.js]: Where is this app used?

File: next.config.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/* eslint-env node */
+
+const withSourceMaps = require('@zeit/next-source-maps')
```
Comment (jorgeorpinel) [inline next.config.js]: I see `@zeit/next-source-maps` was also added here. Is this specifically just a nice addition for Sentry logs/traces?

File: pages/_app.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import React from 'react'
+import App from 'next/app'
+import * as Sentry from '@sentry/browser'
+
+Sentry.init({
+  dsn: process.env.SENTRY_DSN
+})
+
+class MyApp extends App {
```
Comment (jorgeorpinel) [inline pages/_app.js]: Where is this app used?

File: pages/_document.js [hunk 4]
```diff
@@ -1,6 +1,17 @@
...
 import { ServerStyleSheet } from 'styled-components'
+import * as Sentry from '@sentry/browser'
+
+process.on('unhandledRejection', err => {
+  Sentry.captureException(err)
+})
+
+process.on('uncaughtException', err => {
+  Sentry.captureException(err)
+})
```
Comment (jorgeorpinel) [inline pages/_document.js]: Why are these needed if Sentry supposedly has automated handling of uncaught exceptions and unhandled rejections? See https://docs.sentry.io/platforms/javascript/#automatically-capturing-errors

Comment (iAdramelk) [reply pages/_document.js]: It's for server side errors in node.js, Sentry will automatically log errors in browser, but not unhandled errors in SSR.

File: pages/_app.js (hunk 3 above)
Comment (iAdramelk) [reply pages/_app.js]: It's replacement of the standard Next.js component https://nextjs.org/docs/advanced-features/custom-app . Basically it's a wrapper around the pages.

File: next.config.js (hunk 2 above)
Comment (iAdramelk) [reply next.config.js]: Source maps is a mechanism to map bundled and minified code with the original source files. If we just run sentry as it is, it will show errors like: `You nave an error on line 1, symbol 3454560 in the gX method.` Source maps will ideally show us original file, line and method name. (But right now it's looks like it didn't work correctly)

It's not strictly related to Sentry and is used for example by browser Dev Tools, but it make using Sentry more convenient.

File: pages/_document.js (hunk 4 above)
Comment (jorgeorpinel) [reply pages/_document.js]: Got it, thanks.

Would these also be needed separately for server.js or other build-time code? e.g. sidebar.js

File: next.config.js (hunk 2 above)
Comment (jorgeorpinel) [reply next.config.js]: Right, I did read about source maps and it seems like a great feature. I was just wondering if this was added specifically for Sentry traces or whether it can also be used on the Browser now.

Comment (iAdramelk) [reply next.config.js]: Well, we should be, because I leave them enabled (there is also an option to send them directly to sentry through API and not publish on site). But in practice from that I see, I misconfigured them somehow and didn't see them anywhere.

File: pages/_document.js (hunk 4 above)
Comment (iAdramelk) [reply pages/_document.js]: This catches all code in the actual pages including `sidebar.js`, but I think we are now missing errors from server.js because they happens before we call `_document.js`.

File: next.config.js (hunk 2 above)
Comment (jorgeorpinel) [reply next.config.js]: OK cool please let us know when you figure it out 🙂

File: pages/_document.js (hunk 4 above)
Comment (jorgeorpinel) [reply pages/_document.js]: Yes some server code runs before the Next app and partially includes sidebar.js code I think. That's what I was wondering about. But it's probably not a big deal? Only runs when the server is being initialized, probably only when new versions are deployed to Heroku.

Comment (iAdramelk) [reply pages/_document.js]: That's pretty much my thoughts as well. And also the fact that we are planning to migrate to Gatsby and it didn't have a server.  So I decided to leave it as is for now.
